src/util.js

```diff
@@ -624,6 +624,27 @@ function hasNewline(text, index, opts) {
   return idx !== idx2;
 }
 
+/**
+ * Check if string can safely be converted from double to single quotes, i.e.
+ *
+ * - no embedded variables ("foo $bar")
+ * - no linebreaks
+ * - no special characters like \n, \t, ...
+ * - no octal/hex/unicode characters
+ *
+ * See http://php.net/manual/en/language.types.string.php#language.types.string.syntax.double
+ */
+function useSingleQuote(node, options) {
+  return (
+    !node.isDoubleQuote ||
+    (options.singleQuote &&
+      !node.raw.match(/\\n|\\t|\\r|\\t|\\v|\\e|\\f/) &&
+      !node.value.match(
+        /["'$\n]|\\[0-7]{1,3}|\\x[0-9A-Fa-f]{1,2}|\\u{[0-9A-Fa-f]+}/
+      ))
+  );
+}
```

Comment:
I think better avoid using `RegExp`, better traverse `nodes` on only string nodes

Reply:
`useSingleQuote` is only called on String nodes. I moved it into utils because I thought I'd also need to call it from `clean.js` but it seems that `massageASTNode` doesn't have access to the `options` object...

Reply:
@czosel hm, let's fix it in prettier, should be easy

Reply:
Just pushed an alternative implementation that just removes `isDoubleQuote` in `clean.js`. Shall we suggest the change in prettier and refactor in another PR?

Reply:
@czosel yep, let's add TODO

Reply:
@evilebottnawi anything else left to be done?